odoo/models.py

```diff
@@ -6328,29 +6325,32 @@ def _flush(self, fnames=None):
         if not any(field in dirty_fields for field in fields):
             return
 
-        # if any field is context-dependent, the values to flush should
-        # be found with a context where the context keys are all None
-        model = self.with_context(context={})
         id_vals = defaultdict(dict)
         for field in self._fields.values():
             ids = self.env.cache.clear_dirty_field(field)
             if not ids:
                 continue
-            records = model.browse(ids)
-            values = list(self.env.cache.get_values(records, field))
-            assert len(values) == len(records), \
+            values = [
+                value
+                # if any field is context-dependent, the values to flush should
+                # be found with a context where the context keys are all None
+                for value in self.env.cache.get_values(field, None, ids)
+                if value is not api.NOTHING
+            ]
+            assert len(values) == len(ids), \
                 f"Could not find all values of {field} to flush them\n" \
                 f"    Cache: {self.env.cache!r}"
-            for record, value in zip(records, values):
+            for record, value in zip(self.browse(ids), values):
                 if not field.translate:
                     value = field.convert_to_write(value, record)
                     value = field.convert_to_column(value, record)
                 else:
                     value = field._convert_from_cache_to_column(value)
-                id_vals[record.id][field.name] = value
+                fname = f'__{field.name}' if field.translate is True and value and isinstance(value.adapted, odoo.fields.TranslatedCacheValue) else field.name
```

Comment:
user different field name e.g. `__name`
partial translations flush -> use the translations to update jsonb (`||` operator)
full translations flush -> use the translations to replace the jsonb
And for sake of the latest temporary table flush strategy, which allows flushing with less number of quires, partial translations and full translations should be flushed in different ways with different field names here for grouping